[FLINK-38809][Tests] Upgrade Testcontainers to 2.0.3

MartijnVisser opened the pull request:
## What is the purpose of the change

* Update testcontainers to latest version (2.0.3) to avoid Docker API Client compatibility problems

## Brief change log

* Refactored flink-end-to-end-tests-common from JUnit4 to JUnit5. Created local ExternalResource interface to be fully JUnit5-compatible with minimum code changes.
* Updated dependencies and imports

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**yes** / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

MartijnVisser committed:
[FLINK-38809][Tests] Upgrade Testcontainers to 2.0.3

gustavodemorais commented:
Did you test it locally? CI is failing with 
`10:59:29.860 [ERROR] /__w/1/s/flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/cache/LolCache.java:[21,23] package org.junit.rules does not exist`

MartijnVisser committed:
[FLINK-38809][Tests] Migrate flink-end-to-end-tests-common from JUnit4 to JUnit5

Testcontainers 2.0.2 removed JUnit4 support, requiring migration to JUnit5. Created local ExternalResource interface to be fully JUnit5-compatible with minimum code changes.

This migration allows the module to work with Testcontainers 2.0.2 while maintaining all existing test functionality.
